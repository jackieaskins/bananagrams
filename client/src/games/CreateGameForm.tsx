import React from 'react';
import { Grid, Typography } from '@material-ui/core';

import Button from '../buttons/Button';
import ErrorAlert from '../alerts/ErrorAlert';
import TextField from '../forms/TextField';
import { useCreateGameForm } from './CreateGameFormState';

const CreateGameForm: React.FC = () => {
  const {
    error,
    gameName,
    isCreatingGame,
    onSubmit,
    setGameName,
    setUsername,
    username,
  } = useCreateGameForm();

  return (
    <>
      <Typography variant="h3" style={{ marginBottom: '5px' }}>
        Start a new game
      </Typography>
      <ErrorAlert visible={!!error} style={{ marginBottom: '5px' }}>
        {error}
      </ErrorAlert>
      <Grid
        container
        direction="column"
        spacing={3}
        component="form"
        autoComplete="off"
      >
        <Grid item>
          <TextField
            label="Game name"
            name="gameName"
            required
            margin="dense"
            value={gameName}
            setValue={setGameName}
          />

          <TextField
            label="Username"
            name="username"
            margin="dense"
            required
            value={username}
            setValue={setUsername}
          />
        </Grid>

        <Grid item>
          <Button
            loading={isCreatingGame}
            loadingText="Creating game"
            disabled={!gameName || !username}
            onClick={onSubmit}
            size="large"
            fullWidth
          >
            Create game
          </Button>
        </Grid>
      </Grid>
    </>
  );
};

export default CreateGameForm;
