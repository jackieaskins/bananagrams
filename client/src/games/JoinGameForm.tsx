import React from 'react';
import { Grid, Link as MUILink, Typography } from '@material-ui/core';
import { Link } from 'react-router-dom';

import Button from '../buttons/Button';
import ErrorAlert from '../alerts/ErrorAlert';
import TextField from '../forms/TextField';
import { useJoinGameForm } from './JoinGameFormState';

const JoinGameForm: React.FC = () => {
  const {
    error,
    isJoiningGame,
    onSubmit,
    setUsername,
    username,
  } = useJoinGameForm();

  return (
    <>
      <Typography variant="h3" style={{ marginBottom: '5px' }}>
        Join game
      </Typography>
      <ErrorAlert visible={!!error} style={{ marginBottom: '5px' }}>
        {error}
      </ErrorAlert>
      <Grid
        container
        direction="column"
        spacing={3}
        component="form"
        autoComplete="off"
      >
        <Grid item>
          <TextField
            label="Username"
            name="username"
            required
            value={username}
            setValue={setUsername}
          />
        </Grid>

        <Grid item>
          <Button
            disabled={!username}
            loading={isJoiningGame}
            loadingText="Joining game"
            onClick={onSubmit}
            size="large"
            fullWidth
          >
            Join game
          </Button>
        </Grid>

        <Grid item style={{ margin: '0 auto', padding: '0' }}>
          <MUILink component={Link} to="/" variant="body1" align="center">
            Go home
          </MUILink>
        </Grid>
      </Grid>
    </>
  );
};

export default JoinGameForm;
