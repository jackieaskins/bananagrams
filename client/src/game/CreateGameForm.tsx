import { Button, Form as AntForm, Input } from 'antd';
import { useCallback } from 'react';
import { useHistory } from 'react-router-dom';

import Form, { SubmitFormFn } from '../form/Form';
import { GameInfo, GameLocationState } from '../games/types';
import { useSocket } from '../socket/SocketContext';

const FormItem = AntForm.Item;

type FormValues = {
  gameName: string;
  username: string;
};

type CreateGameFormProps = {
  isShortenedGame?: boolean;
};

const CreateGameForm = ({
  isShortenedGame = false,
}: CreateGameFormProps): JSX.Element => {
  const { socket } = useSocket();
  const { push } = useHistory();

  const handleSubmit: SubmitFormFn<FormValues> = useCallback(
    (setError) => ({ gameName, username }) => {
      socket.emit(
        'createGame',
        { gameName, username, isShortenedGame },
        (error: Error, gameInfo: GameInfo) => {
          if (error) {
            setError(error.message);
          } else {
            const locationState: GameLocationState = {
              isInGame: true,
              gameInfo,
            };

            push(`/game/${gameInfo.gameId}`, locationState);
          }
        }
      );
    },
    [isShortenedGame, push, socket]
  );

  return (
    <Form labelCol={{ span: 5 }} onSubmit={handleSubmit}>
      <FormItem label="Game name" name="gameName" rules={[{ required: true }]}>
        <Input />
      </FormItem>

      <FormItem label="Username" name="username" rules={[{ required: true }]}>
        <Input />
      </FormItem>

      <FormItem>
        <Button type="primary" htmlType="submit" block>
          Create game
        </Button>
      </FormItem>
    </Form>
  );
};

export default CreateGameForm;
