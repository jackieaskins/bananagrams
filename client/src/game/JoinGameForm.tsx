import { Form as AntForm, Input, Button } from 'antd';
import { useCallback } from 'react';
import { useHistory } from 'react-router-dom';

import Form, { SubmitFormFn } from '../form/Form';
import { GameInfo, GameLocationState } from '../games/types';
import { useSocket } from '../socket/SocketContext';

const FormItem = AntForm.Item;

type FormValues = {
  gameId: string;
  username: string;
};

type JoinGameFormProps = {
  gameId: string;
  showGameIdField?: boolean;
};

const JoinGameForm = ({
  gameId,
  showGameIdField = true,
}: JoinGameFormProps): JSX.Element => {
  const { socket } = useSocket();
  const { push } = useHistory();

  const handleSubmit: SubmitFormFn<FormValues> = useCallback(
    (setError) => ({ gameId, username }) => {
      socket.emit(
        'joinGame',
        { gameId, username },
        (error: Error, gameInfo: GameInfo) => {
          if (error) {
            setError(error.message);
          } else {
            const locationState: GameLocationState = {
              isInGame: true,
              gameInfo,
            };

            push(`/game/${gameId}`, locationState);
          }
        }
      );
    },
    [push, socket]
  );

  return (
    <Form
      labelCol={{ span: 5 }}
      onSubmit={handleSubmit}
      initialValues={{ gameId }}
    >
      <FormItem
        hidden={!showGameIdField}
        label="Game id"
        name="gameId"
        rules={[{ required: true }]}
      >
        <Input />
      </FormItem>

      <FormItem label="Username" name="username" rules={[{ required: true }]}>
        <Input />
      </FormItem>

      <FormItem>
        <Button type="primary" htmlType="submit" block>
          Join game
        </Button>
      </FormItem>
    </Form>
  );
};

export default JoinGameForm;
