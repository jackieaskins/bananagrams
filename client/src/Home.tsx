import { Card, Col, Tabs } from 'antd';
import { Row } from 'antd/lib/grid';
import { useCallback, useState } from 'react';
import { useHistory, useLocation } from 'react-router-dom';

import CreateGameForm from './game/CreateGameForm';
import JoinGameForm from './game/JoinGameForm';

enum TabKey {
  CreateGame = 'createGame',
  JoinGame = 'joinGame',
}

const Home = (): JSX.Element => {
  const { search } = useLocation();
  const { replace } = useHistory();

  const params = new URLSearchParams(search);
  const { gameId, tabKey } = Object.fromEntries(params);
  const isShortenedGame = params.has('shortenedGame');

  const [activeTabKey, setActiveTabKey] = useState(
    Object.values(TabKey).includes(tabKey as TabKey)
      ? tabKey
      : TabKey.CreateGame
  );

  const handleTabChange = useCallback(
    (key) => {
      const queryParams = new URLSearchParams(search);
      queryParams.set('tabKey', key);

      setActiveTabKey(key);
      replace({ search: queryParams.toString() });
    },
    [replace, search]
  );

  return (
    <Row css={{ height: '100vh' }} align="middle" justify="center">
      <Col xs={23} sm={21} md={16} lg={12}>
        <Card>
          <h1 css={{ textAlign: 'center' }}>Welcome to Bananagrams!</h1>
          <Tabs
            activeKey={activeTabKey}
            centered
            onChange={handleTabChange}
            size="small"
          >
            <Tabs.TabPane key={TabKey.CreateGame} tab="Create game">
              <CreateGameForm isShortenedGame={isShortenedGame} />
            </Tabs.TabPane>

            <Tabs.TabPane key={TabKey.JoinGame} tab="Join game">
              <JoinGameForm gameId={gameId} />
            </Tabs.TabPane>
          </Tabs>
        </Card>
      </Col>
    </Row>
  );
};

export default Home;
