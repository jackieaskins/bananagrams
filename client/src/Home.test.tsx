import { Tabs } from 'antd';
import { shallow } from 'enzyme';
import { useLocation } from 'react-router-dom';

import Home from './Home';
import CreateGameForm from './game/CreateGameForm';
import JoinGameForm from './game/JoinGameForm';

const mockSetActiveTabKey = jest.fn();
jest.mock('react', () => ({
  ...jest.requireActual<any>('react'),
  useState: jest.fn().mockImplementation((init) => [init, mockSetActiveTabKey]),
}));

const mockReplace = jest.fn();
jest.mock('react-router-dom', () => ({
  useHistory: () => ({ replace: mockReplace }),
  useLocation: jest.fn(),
}));

const mockUseLocation = useLocation as jest.Mock;

describe('<Home />', () => {
  beforeEach(() => {
    mockUseLocation.mockReturnValue({ search: '' });
  });

  test('renders centered form tabs card', () => {
    expect(shallow(<Home />)).toMatchSnapshot();
  });

  test('initializes activeTabKey if query param is present', () => {
    mockUseLocation.mockReturnValue({ search: '?tabKey=joinGame' });

    expect(
      shallow(<Home />)
        .find(Tabs)
        .props().activeKey
    ).toEqual('joinGame');
  });

  test('passes isShortenedGame to CreateGameForm', () => {
    mockUseLocation.mockReturnValue({ search: '?shortenedGame' });

    expect(
      shallow(<Home />)
        .find(CreateGameForm)
        .props().isShortenedGame
    ).toEqual(true);
  });

  test('passes gameId to JoinGameForm', () => {
    mockUseLocation.mockReturnValue({ search: '?gameId=gameId' });

    expect(
      shallow(<Home />)
        .find(JoinGameForm)
        .props().gameId
    ).toEqual('gameId');
  });

  describe('onTabChange', () => {
    const newKey = 'joinGame';

    beforeEach(() => {
      shallow(<Home />)
        .find(Tabs)
        .props()
        .onChange?.(newKey);
    });

    test('sets active tab key', () => {
      expect(mockSetActiveTabKey).toHaveBeenCalledWith(newKey);
    });

    test('adds query param to url', () => {
      expect(mockReplace).toHaveBeenCalledWith({ search: `tabKey=${newKey}` });
    });
  });
});
