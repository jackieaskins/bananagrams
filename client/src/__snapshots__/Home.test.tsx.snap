// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders centered form tabs card 1`] = `
.emotion-0awwaht-Home {
  height: 100vh;
}

.emotion-05hhcs4-Home {
  text-align: center;
}

<undefined
  align="middle"
  className="emotion-0"
  justify="center"
>
  <Col
    lg={12}
    md={16}
    sm={21}
    xs={23}
  >
    <Card>
      <h1
        className="emotion-0"
      >
        Welcome to Bananagrams!
      </h1>
      <Tabs
        activeKey="createGame"
        centered={true}
        onChange={[Function]}
        size="small"
      >
        <TabPane
          key="createGame"
          tab="Create game"
        >
          <CreateGameForm
            isShortenedGame={false}
          />
        </TabPane>
        <TabPane
          key="joinGame"
          tab="Join game"
        >
          <JoinGameForm />
        </TabPane>
      </Tabs>
    </Card>
  </Col>
</undefined>
`;
